### ISSUE TITLE
Rewrite iocpreactor using code taken from twisted.internet.{tcp|abstract|udp}

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#1760]
|<img alt="PenguinOfDoom's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| PenguinOfDoom reported|
|-|-|
|Trac ID|trac#1760|
|Type|enhancement|
|Created|2006-05-29 23:54:00Z|
|Branch|https://github.com/twisted/twisted/tree/iocpreactor-1760-6|

By carefully monkeypatching select, socket or doSelect (or parameterizing), one could reuse most of twisted.internet for alternative concurrency mechanisms.

Attachments:

* [1760.zip](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/94f/94fe6a3196c44b2cd7c2ea7776add10deb1fd968/6b6b14d27152ffa81078413501d39ad8e5507be2.zip) (2223 bytes) - added by ghazel on 2006-12-07 21:57:32Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__1760 1760
type__enhancement enhancement
reporter__PenguinOfDoom PenguinOfDoom
priority__normal normal
milestone__ 
branch__branches_iocpreactor_1760_6 branches/iocpreactor-1760-6
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1148946840000000 1148946840000000
changetime__1205861762000000 1205861762000000
version__None None
owner__ 
cc__exarkun cc__ghazel cc__itamarst
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun commented|
|-|-|

I think you typed "monkeypatching" when you meant to type "refactoring".

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun commented|
|-|-|

How's this going?

Have we done any testing of real applications with this code yet?  Is it ready for that?

**Comment by twisted-trac** [comment]
|[<img alt="ghazel's avatar" src="https://avatars.githubusercontent.com/u/31402?s=50" width="50" height="50">](https://github.com/ghazel)<a name="note_4"></a>|@ghazel commented|
|-|-|

I gave iocpreactor-1760-3 a spin, and ran in to a few issues:

- build.bat specifies mingw32. I took a stab and removed that in favor of msvc7.1 which python was built with

- it couldn't find python24.lib. I specified c:\python24\libs\ in library_dirs and it compiled

- t.i.iocp.reactor._doIteration fails when timeout is None (pretty often). Pulling a selectreactor and doing "if timeout is None: timeout = 0.01" got that running

- memory usage is an order of magnitude (21mB vs 119mB with 800 local to local connections) better than iocp1 or 2, but still not as small as selectreactor (15mB) or my growing buffersize patch for iocp1 (15mB). This is livable, I just thought I'd mention it.

- using pause/resumeProducing produced this error:

```
Traceback (most recent call last):
  File "twisted\python\log.py", line 48, in callWithLogger
    return callWithContext({"system": lp}, func, *args, **kw)
  File "twisted\python\log.py", line 33, in callWithContext
    return context.call({ILogContext: newCtx}, func, *args, **kw)
  File "twisted\python\context.py", line 59, in callWithContext
    return self.currentContext().callWithContext(ctx, func, *args, **kw)
  File "twisted\python\context.py", line 37, in callWithContext
    return func(*args,**kw)
--- <exception caught here> ---
  File "C:\Documents and Settings\user\Desktop\projects\twisted\Twisted\branches
\iocpreactor-1760-3\twisted\internet\iocpreactor\reactor.py", line 106, in _call
EventCallback
    evt.callback(rc, bytes, evt)
  File "twisted\internet\iocpreactor\abstract.py", line 70, in _cbRead
    if self._handleRead(rc, bytes, evt):
  File "twisted\internet\iocpreactor\abstract.py", line 84, in _handleRead
    assert self._readSize == 0, str(self._readSize) # you're welcome
exceptions.AssertionError: 4096
```
just before that, _dispatchData returned from inside the "while self._readNextBuffer < full_buffers" loop, presumably because producing was paused during the loop.

- python24.dll often crashes just after that, with an access violation.

**Comment by twisted-trac** [comment]
|<img alt="PenguinOfDoom's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_5"></a>|PenguinOfDoom commented|
|-|-|

- Don't use build.bat. It's not for you. Use setup.py in the branch root.
- distutils finds Python headers and libraries automatically. Perhaps your install is broken.
- In what situation is doIteration called with timeout of None?
- Perhaps it might be worthwhile to decrease FileHandle.readBufferSize to 1024 and bump FileHandle.maxReadBuffers to 64, or whatever. I prefer to waste a small amount of memory per connection and avoid constantly allocating and deallocating arbitrarily-sized buffers.
- I would greatly appreciate a minimal example that produces this behavior.

**Comment by twisted-trac** [comment]
|[<img alt="ghazel's avatar" src="https://avatars.githubusercontent.com/u/31402?s=50" width="50" height="50">](https://github.com/ghazel)<a name="note_6"></a>|@ghazel commented|
|-|-|

doIteration is called with a timeout of None in this simple example:
```
from twisted.internet import iocpreactor
iocpreactor.install()
from twisted.internet import reactor
reactor.run()
```

I added 1760.zip, an example that reproduces the producer problem. Basically, it's an example that forms 5 connections, and spews data across them. The download side has a rate limiter, and because of all that pause/resume producing the bug is displayed.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_7"></a>|@itamarst commented|
|-|-|

http://www.summerblue.net/computing/libiocp/index.html may or may not be helpful.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun commented|
|-|-|

I merged this branch forward and resolved some conflicts in it.  Some of the non-iocp changes should probably be broken out into separate tickets.

The code is in iocpreactor-1760-4 for the moment, though.  Note that svn 1.4 is required to merge this branch correctly.

**Comment by twisted-trac** [comment]
|<img alt="pahan's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_9"></a>|pahan commented|
|-|-|

(In [22833]) Branching to 'iocpreactor-1760-5'

**Comment by twisted-trac** [comment]
|<img alt="pahan's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_10"></a>|pahan commented|
|-|-|

(In [22838]) Branching to 'iocpreactor-1760-6'

**Comment by twisted-trac** [comment]
|<img alt="PenguinOfDoom's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_11"></a>|PenguinOfDoom set owner to @therve|
|-|-|

It passes all tests

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_12"></a>|@therve set owner to PenguinOfDoom|
|-|-|

 * reactor.py uses a set(), it should use a twisted.python.compat.set object
 * the way the reactor attribute of connection objects is managed is wrong. It should be done at the lowest level by `FileHandle`, like what `FileDescriptor` is doing.
 * the default timeout of 0 in doIterate in wrong I think. 1 is probably a more reasonable default value. If you just call reactor.run(), it eats all the CPU.

I can probably nitpicking forever, but I think it's good now. So do these little things, run the tests suite, and then please merge.

Thanks.

**Comment by twisted-trac** [comment]
|<img alt="pahan's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_13"></a>|pahan set status to `closed`|
|-|-|

(In [22957]) Merge iocpreactor-1760-6: new iocpreactor implementation

This is new and improved iocpreactor that passes all tests.

Author: PenguinOfDoom
Reviewer: therve
Fixes [#1760](https://github.com/twisted/twisted/issues/1760)

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
